src/rust/ide/enso-protocol/src/file_manager.rs

```diff
@@ -0,0 +1,411 @@
+        executor.spawner().spawn_local(client.runner()).unwrap();
+        Fixture {transport,client,executor}
+    }
+
+    #[test]
+    fn test_notification() {
+        let mut fixture = setup_fm();
+        let mut events  = Box::pin(fixture.client.events());
+        assert!(poll_stream_output(&mut events).is_none());
+
+        let expected_notification = FilesystemEvent {
+            path : Path::new("./Main.luna"),
+            kind : FilesystemEventKind::Modified,
+        };
+        let notification_text = r#"{
+            "jsonrpc": "2.0",
+            "method": "filesystemEvent",
+            "params": {"path" : "./Main.luna", "kind" : "Modified"}
+        }"#;
+        fixture.transport.mock_peer_message_text(notification_text);
+        assert!(poll_stream_output(&mut events).is_none());
+
+        fixture.executor.run_until_stalled();
+
+        let event = poll_stream_output(&mut events);
+        if let Some(Event::Notification(n)) = event {
+            assert_eq!(n, Notification::FilesystemEvent(expected_notification));
+        } else {
+            panic!("expected notification event");
+        }
+    }
+
+    /// Tests making a request using file manager:
+    /// * creates FM client and uses `make_request` to make a request
+    /// * checks that request is made for `expected_method`
+    /// * checks that request input is `expected_input`
+    /// * mocks receiving a response from server with `result`
+    /// * checks that FM-returned Future yields `expected_output`
+    fn test_request<Fun, Fut, T>
+    ( make_request:Fun
+    , expected_method:&str
+    , expected_input:Value
+    , result:Value
+    , expected_output:T )
+    where Fun : FnOnce(&mut Client) -> Fut,
+          Fut : Future<Output = Result<T>>,
+          T   : Debug + PartialEq {
+        let mut fixture = setup_fm();
+        let mut fut     = Box::pin(make_request(&mut fixture.client));
+
+        let request = fixture.transport.expect_message::<RequestMessage<Value>>();
+        assert_eq!(request.method, expected_method);
+        assert_eq!(request.params, expected_input);
+
+        let response = Message::new_success(request.id, result);
+        fixture.transport.mock_peer_message(response);
+        fixture.executor.run_until_stalled();
+        let output = poll_future_output(&mut fut).unwrap().unwrap();
+        assert_eq!(output, expected_output);
+    }
+
+    #[test]
+    fn test_requests() {
+        let main                = Path::new("./Main.luna");
+        let target              = Path::new("./Target.luna");
+        let path_main           = json!({"path" : "./Main.luna"});
+        let from_main_to_target = json!({
+            "from" : "./Main.luna",
+            "to"   : "./Target.luna"
+        });
+        let true_json = json!(true);
+        let unit_json = json!(null);
+
+        test_request(
+            |client| client.copy_directory(main.clone(), target.clone()),
```

Comment:
why the API requries me to pass it with ownership of path? This is very surprising. All APIs like that in Rust libs would definitely be written not requiring it. We should use the API as `client.copy_directory(&main,&target)` instead.

Reply:
I do not see this neither ficxed nor I see FIXME in the code regarding it. The FIXME could lead to the newly created issue about it.